File: src/utils/names.ts
/**
 * Name Generator Module
 * 
 * Handles first name and surname selection based on the resolved
 * demographic path. Uses religion, state, caste, and gender context
 * to pick culturally accurate names.
 */

import type { SeededRNG, CompiledDatabase, Gender, NameEntry } from '../types.js';
import { weightedSample, uniformSample } from '../core/sampler.js';

function getStateRegion(stateId: string): string {
  const south = ['karnataka', 'kerala', 'tamil_nadu', 'andhra_pradesh', 'telangana', 'puducherry', 'lakshadweep'];
  const east = ['west_bengal', 'odisha', 'bihar', 'jharkhand', 'assam', 'sikkim', 'arunachal_pradesh', 'nagaland', 'manipur', 'mizoram', 'tripura', 'meghalaya'];
  const north = ['punjab', 'haryana', 'delhi', 'himachal_pradesh', 'jammu_kashmir', 'uttarakhand', 'chandigarh', 'ladakh'];
  const west = ['maharashtra', 'gujarat', 'goa', 'rajasthan', 'dadra_nagar_haveli', 'daman_diu'];

  if (south.includes(stateId)) return 'south';
  if (east.includes(stateId)) return 'east';
  if (north.includes(stateId)) return 'north';
  if (west.includes(stateId)) return 'west';
  return 'default';
}

/**
 * Select a first name based on religion, state, and gender.
 * Cascading lookup: religion+state → religion+region → religion+default → any default
 */
export function selectFirstName(
  db: CompiledDatabase,
  religionId: string,
  stateId: string,
  gender: Gender,
  rng: SeededRNG
): { name: string; probability: number } {
  const byReligion = db.firstNames[religionId];
  
  if (byReligion) {
    // Try state-specific names first
    let byState = byReligion[stateId];
    if (!byState) {
      // Try regional grouping fallback
      const regionId = getStateRegion(stateId);
      byState = byReligion[regionId];
    }
    
    if (byState && byState[gender] && byState[gender].length > 0) {
      return sampleFromNameList(byState[gender], rng);
    }
    
    // Fallback to default names for this religion
    const byDefault = byReligion['default'];
    if (byDefault && byDefault[gender] && byDefault[gender].length > 0) {
      return sampleFromNameList(byDefault[gender], rng);
    }
    
    // Try 'other' gender
    if (byDefault && byDefault['other'] && byDefault['other'].length > 0) {
      return sampleFromNameList(byDefault['other'], rng);
    }
  }
  
  // Last resort: pick from Hindu defaults (most common)
  const hinduDefault = db.firstNames['hindu']?.['default']?.[gender];
  if (hinduDefault && hinduDefault.length > 0) {
    return sampleFromNameList(hinduDefault, rng);
  }
  
  return { name: gender === 'male' ? 'Arjun' : 'Priya', probability: 0.01 };
}

/**
 * Select a surname based on caste/community.
 * Cascading lookup: casteId → normalized casteId → generic
 */
export function selectSurname(
  db: CompiledDatabase,
  casteId: string,
  gender: Gender,
  rng: SeededRNG
): { name: string; probability: number } {
  // Direct caste match
  let surnameList = db.surnames[casteId];
  
  if (!surnameList || surnameList.length === 0) {
    // Try normalized key
    const normalizedKey = casteId.toLowerCase().replace(/[\s\-]+/g, '_');
    surnameList = db.surnames[normalizedKey];
  }
  
  if (!surnameList || surnameList.length === 0) {
    // Try partial match
    for (const [key, list] of Object.entries(db.surnames)) {
      if (key.includes(casteId) || casteId.includes(key)) {
        surnameList = list;
        break;
      }
    }
  }
  
  if (surnameList && surnameList.length > 0) {
    // Filter by gender if applicable (some surnames are gender-specific, e.g., Sikh: Singh/Kaur)
    const genderFiltered = surnameList.filter(s => s.gender === gender || s.gender === 'unisex');
    if (genderFiltered.length > 0) {
      return sampleFromNameList(genderFiltered, rng);
    }
    return sampleFromNameList(surnameList, rng);
  }
  
  // Generic fallback surnames
  const genericSurnames: NameEntry[] = [
    { name: 'Kumar', weight: 15, gender: 'male' },
    { name: 'Kumari', weight: 10, gender: 'female' },
    { name: 'Devi', weight: 10, gender: 'female' },
    { name: 'Singh', weight: 12, gender: 'male' },
    { name: 'Prasad', weight: 8, gender: 'male' },
    { name: 'Das', weight: 8, gender: 'unisex' },
    { name: 'Ram', weight: 5, gender: 'male' },
    { name: 'Lal', weight: 5, gender: 'male' }
  ];
  
  const filtered = genericSurnames.filter(s => s.gender === gender || s.gender === 'unisex');
  if (filtered.length > 0) {
    return sampleFromNameList(filtered, rng);
  }
  return sampleFromNameList(genericSurnames, rng);
}

/**
 * Select a mother tongue based on state's language distribution.
 */
export function selectMotherTongue(
  db: CompiledDatabase,
  stateId: string,
  rng: SeededRNG
): string {
  const stateData = db.states[stateId];
  if (!stateData || !stateData.languageDistribution) {
    return 'Hindi';
  }
  
  const entries = Object.entries(stateData.languageDistribution);
  if (entries.length === 0) return 'Hindi';
  
  const items = entries.map(([lang, weight]) => ({ name: lang, weight }));
  const { item } = weightedSample(items, rng);
  
  // Capitalize first letter
  return item.name.charAt(0).toUpperCase() + item.name.slice(1);
}

/**
 * Select a second language (different from mother tongue).
 */
export function selectSecondLanguage(
  db: CompiledDatabase,
  stateId: string,
  motherTongue: string,
  education: string,
  rng: SeededRNG
): string | undefined {
  // Lower-educated people are less likely to know a second language
  const eduLevels = ['illiterate', 'literate_below_primary', 'primary'];
  if (eduLevels.includes(education)) {
    if (rng.next() > 0.3) return undefined;
  }
  
  // Common second languages
  const secondLanguages = [
    { name: 'Hindi', weight: 30 },
    { name: 'English', weight: 25 },
    { name: 'Urdu', weight: 5 },
    { name: 'Bengali', weight: 3 },
    { name: 'Tamil', weight: 3 },
    { name: 'Telugu', weight: 3 },
    { name: 'Marathi', weight: 3 },
    { name: 'Gujarati', weight: 2 },
    { name: 'Kannada', weight: 2 },
    { name: 'Malayalam', weight: 2 },
    { name: 'Punjabi', weight: 2 },
    { name: 'Odia', weight: 1 }
  ];
  
  // Filter out mother tongue
  const mtLower = motherTongue.toLowerCase();
  const available = secondLanguages.filter(l => l.name.toLowerCase() !== mtLower);
  
  if (available.length === 0) return undefined;
  
  const { item } = weightedSample(available, rng);
  return item.name;
}

/**
 * Select a district based on state.
 */
export function selectDistrict(
  db: CompiledDatabase,
  stateId: string,
  rng: SeededRNG
): string {
  const districtList = db.districts[stateId];
  if (!districtList || districtList.length === 0) {
    return db.states[stateId]?.stateName ?? 'Unknown';
  }
  return uniformSample(districtList, rng);
}

// ─────────────────────────────────────────────────────────────
// Internal Helpers
// ─────────────────────────────────────────────────────────────

function sampleFromNameList(
  names: NameEntry[],
  rng: SeededRNG
): { name: string; probability: number } {
  const { item, probability } = weightedSample(names, rng);
  return { name: item.name, probability };
}

File: python/src/indian_fakedata/utils/names.py
"""
Name Generator Module

Handles first name and surname selection based on the resolved
demographic path. Uses religion, state, caste, and gender context
to pick culturally accurate names.
"""

import re
from indian_fakedata.core.sampler import weighted_sample, uniform_sample


def _get_state_region(state_id):
    """Map a state to its cultural region."""
    south = ['karnataka', 'kerala', 'tamil_nadu', 'andhra_pradesh', 'telangana', 'puducherry', 'lakshadweep']
    east = ['west_bengal', 'odisha', 'bihar', 'jharkhand', 'assam', 'sikkim',
            'arunachal_pradesh', 'nagaland', 'manipur', 'mizoram', 'tripura', 'meghalaya']
    north = ['punjab', 'haryana', 'delhi', 'himachal_pradesh', 'jammu_kashmir', 'uttarakhand', 'chandigarh', 'ladakh']
    west = ['maharashtra', 'gujarat', 'goa', 'rajasthan', 'dadra_nagar_haveli', 'daman_diu']

    if state_id in south:
        return 'south'
    if state_id in east:
        return 'east'
    if state_id in north:
        return 'north'
    if state_id in west:
        return 'west'
    return 'default'


def _sample_from_name_list(names, rng):
    """Sample a name from a NameEntry list, returns (name, probability)."""
    item, probability = weighted_sample(names, rng)
    name_str = item.get("name", "Unknown") if isinstance(item, dict) else getattr(item, "name", "Unknown")
    return name_str, probability


def select_first_name(db, religion_id, state_id, gender, rng):
    """
    Select a first name based on religion, state, and gender.
    Cascading lookup: religion+state → religion+region → religion+default → any default
    """
    by_religion = db.get("firstNames", {}).get(religion_id)

    if by_religion:
        # Try state-specific names first
        by_state = by_religion.get(state_id)
        if not by_state:
            region_id = _get_state_region(state_id)
            by_state = by_religion.get(region_id)

        if by_state and by_state.get(gender) and len(by_state[gender]) > 0:
            return _sample_from_name_list(by_state[gender], rng)

        # Fallback to default names for this religion
        by_default = by_religion.get('default')
        if by_default and by_default.get(gender) and len(by_default[gender]) > 0:
            return _sample_from_name_list(by_default[gender], rng)

        # Try 'other' gender
        if by_default and by_default.get('other') and len(by_default['other']) > 0:
            return _sample_from_name_list(by_default['other'], rng)

    # Last resort: pick from Hindu defaults (most common)
    hindu_default = db.get("firstNames", {}).get('hindu', {}).get('default', {}).get(gender)
    if hindu_default and len(hindu_default) > 0:
        return _sample_from_name_list(hindu_default, rng)

    return ('Arjun' if gender == 'male' else 'Priya', 0.01)


def select_surname(db, caste_id, gender, rng):
    """
    Select a surname based on caste/community.
    Cascading lookup: casteId → normalized casteId → generic
    """
    surnames_db = db.get("surnames", {})

    # Direct caste match
    surname_list = surnames_db.get(caste_id)

    if not surname_list or len(surname_list) == 0:
        # Try normalized key
        normalized_key = re.sub(r'[\s\-]+', '_', caste_id.lower())
        surname_list = surnames_db.get(normalized_key)

    if not surname_list or len(surname_list) == 0:
        # Try partial match
        for key, lst in surnames_db.items():
            if key in caste_id or caste_id in key:
                surname_list = lst
                break

    if surname_list and len(surname_list) > 0:
        # Filter by gender if applicable (e.g., Sikh: Singh/Kaur)
        gender_filtered = [s for s in surname_list
                           if s.get("gender") == gender or s.get("gender") == "unisex"]
        if len(gender_filtered) > 0:
            return _sample_from_name_list(gender_filtered, rng)
        return _sample_from_name_list(surname_list, rng)

    # Generic fallback surnames
    generic_surnames = [
        {"name": "Kumar", "weight": 15, "gender": "male"},
        {"name": "Kumari", "weight": 10, "gender": "female"},
        {"name": "Devi", "weight": 10, "gender": "female"},
        {"name": "Singh", "weight": 12, "gender": "male"},
        {"name": "Prasad", "weight": 8, "gender": "male"},
        {"name": "Das", "weight": 8, "gender": "unisex"},
        {"name": "Ram", "weight": 5, "gender": "male"},
        {"name": "Lal", "weight": 5, "gender": "male"},
    ]

    filtered = [s for s in generic_surnames if s["gender"] == gender or s["gender"] == "unisex"]
    if len(filtered) > 0:
        return _sample_from_name_list(filtered, rng)
    return _sample_from_name_list(generic_surnames, rng)


def select_mother_tongue(db, state_id, rng):
    """Select a mother tongue based on state's language distribution."""
    state_data = db.get("states", {}).get(state_id)
    if not state_data or not state_data.get("languageDistribution"):
        return "Hindi"

    lang_dist = state_data["languageDistribution"]
    if len(lang_dist) == 0:
        return "Hindi"

    items = [{"name": lang, "weight": w} for lang, w in lang_dist.items()]
    item, _ = weighted_sample(items, rng)
    name = item["name"]
    return name[0].upper() + name[1:] if name else "Hindi"


def select_second_language(db, state_id, mother_tongue, education, rng):
    """Select a second language (different from mother tongue)."""
    # Lower-educated people are less likely to know a second language
    low_edu = ['illiterate', 'literate_below_primary', 'primary']
    if education in low_edu:
        if rng.next() > 0.3:
            return None

    second_languages = [
        {"name": "Hindi", "weight": 30},
        {"name": "English", "weight": 25},
        {"name": "Urdu", "weight": 5},
        {"name": "Bengali", "weight": 3},
        {"name": "Tamil", "weight": 3},
        {"name": "Telugu", "weight": 3},
        {"name": "Marathi", "weight": 3},
        {"name": "Gujarati", "weight": 2},
        {"name": "Kannada", "weight": 2},
        {"name": "Malayalam", "weight": 2},
        {"name": "Punjabi", "weight": 2},
        {"name": "Odia", "weight": 1},
    ]

    mt_lower = mother_tongue.lower()
    available = [l for l in second_languages if l["name"].lower() != mt_lower]

    if len(available) == 0:
        return None

    item, _ = weighted_sample(available, rng)
    return item["name"]


def select_district(db, state_id, rng):
    """Select a district based on state."""
    district_list = db.get("districts", {}).get(state_id)
    if not district_list or len(district_list) == 0:
        state_data = db.get("states", {}).get(state_id, {})
        return state_data.get("stateName", "Unknown")
    return uniform_sample(district_list, rng)
